packages/jupyter-ai/jupyter_ai/chat_handlers/base.py

```diff
@@ -1,35 +1,83 @@
-from typing import TYPE_CHECKING, Any, Dict, Optional, Type
+from typing import TYPE_CHECKING, Awaitable, ClassVar, Dict, List, Optional, Type
 from uuid import uuid4
 
+from dask.distributed import Client as DaskClient
 from jupyter_ai.config_manager import ConfigManager, Logger
-from jupyter_ai.models import AgentChatMessage, HumanChatMessage
+from jupyter_ai.models import AgentChatMessage, ChatMessage, HumanChatMessage
 from jupyter_ai_magics.providers import BaseProvider
 
+# necessary to prevent circular import
+from pydantic import BaseModel
+from traitlets.config import Configurable
+
 if TYPE_CHECKING:
     from jupyter_ai.handlers import RootChatHandler
 
 
-class BaseChatHandler:
+# Chat handler type, with specific attributes for each
+class HandlerRoutingType(BaseModel):
+    routing_method: ClassVar[str] = ...
+    """The routing method that sends commands to this handler.
+    Either "natural_language" or "slash_command"."""
+
+
+class SlashCommandRoutingType(HandlerRoutingType):
+    routing_method = "slash_command"
+
+    slash_id: Optional[str]
+    """Slash ID for routing a chat command to this handler. Only one handler
+    may declare a particular slash ID. Must contain only alphanumerics and
+    underscores."""
+
+
+class NaturalLanguageRoutingType(HandlerRoutingType):
+    routing_method = "natural_language"
+
+    description: str
+    """Description used for routing requests, to be used when dispatching
+    messages to model providers. Also shown in the UI for transparency;
+    optimized for model interpretation, not human-facing help."""
+
+
+class BaseChatHandler(Configurable):
     """Base ChatHandler class containing shared methods and attributes used by
     multiple chat handler classes."""
 
+    # Class attributes
+    id: ClassVar[str] = ...
+    """ID for this chat handler; should be unique"""
+
+    name: ClassVar[str] = ...
+    """User-facing name of this handler"""
+
+    help: ClassVar[str] = ...
+    """What this chat handler does, which third-party models it contacts,
+    the format of the data it returns to the user, etc. Used in the UI."""
+
+    routing_type: HandlerRoutingType = ...
+
     def __init__(
         self,
         log: Logger,
         config_manager: ConfigManager,
         root_chat_handlers: Dict[str, "RootChatHandler"],
         model_parameters: Dict[str, Dict],
+        chat_history: List[ChatMessage],
+        root_dir: str,
+        dask_client_future: Awaitable[DaskClient],
     ):
         self.log = log
         self.config_manager = config_manager
         self._root_chat_handlers = root_chat_handlers
         self.model_parameters = model_parameters
+        self._chat_history = chat_history
         self.parser = argparse.ArgumentParser()
+        self.root_dir = os.path.abspath(os.path.expanduser(root_dir))
+        self.dask_client_future = dask_client_future
         self.llm = None
         self.llm_params = None
         self.llm_chain = None
```

Comment:
Calling `__init__` on a traitlets config object is a bit of an anti-pattern. 

Ideally, all of these attributes should be defined as traits and their default values set statically by the trait, or dynamically by `@default` method. 

Fixing this might be beyond the scope of this PR; however, since inheriting this class is how users define custom chat handlers, it might be worth addressing this earlier rather later to avoid subclasses with this pattern.

Reply:
@Zsailer After discussing with my team, I've removed the newly-added inheritance from `Configurable`, since there weren't any traitlets in this class. (There were attributes using `typing` types, but as I learned this afternoon, `traitlets` doesn't play nicely with `typing`, particularly not with `Union` types, which we had been using all over the place.)

I've updated this PR; would you mind taking another look?

Reply:
I'm fine with this change. Since you weren't using traits in this class, you don't lose anything. 

That said, I think there's a strong case for y'all to use traitlets across the library.

First, using traitlets or pydantic models for type validation would improve the developer experience for people wanted to inherit and develop their own chat handler. I would love to see this move towards a system that goes beyond type hints/annotations and towards type validation.

Second, if you leaned into using traitlets, you could consolidate the logging, config_manager, type validation for all objects under a single class, `LoggingConfigurable`. For example, in this class, the config_manager, log, and type validation could easily be shared across all classes/handlers.

> traitlets doesn't play nicely with typing, particularly not with Union types

I'd be curious to explore this point more. A lot of work has been done recently to improve the typing/traitlets experience. There might be an answer to this. 

A lot of folks (including me) like to give traitlets a hard time, but it is pretty magical (sometimes too magical). When used correctly, it can really reduce the complexity of your codebase by shouldering the config, logging, and validation logic.